kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java

```diff
@@ -20,104 +35,319 @@
+    }
+  }
+
+  private static long getDateDiffMinutes(Date date1, Date date2, TimeUnit timeUnit) {
+    long diffInMillies = date2.getTime() - date1.getTime();
+    return timeUnit.convert(diffInMillies, TimeUnit.MILLISECONDS);
+  }
+
+  @Nullable
+  private Descriptor getDescriptor(String fullyQualifiedTypeName) {
+    Date currentDate = new Date();
+    CachedDescriptor cachedDescriptor = cachedMssageDescriptorMap.get(fullyQualifiedTypeName);
+    if (cachedDescriptor != null) {
+      if (getDateDiffMinutes(cachedDescriptor.getTimeCached(), currentDate, TimeUnit.SECONDS)
+          < cachedMessageDescriptorRetensionSeconds) {
+        return cachedDescriptor.getDescriptor();
+      }
+    }
+
+    String bufOwner = bufDefaultOwner;
+    String bufRepo = "";
+
+    String bufOwnerRepoInfo = bufOwnerRepoByProtobufMessageName.get(fullyQualifiedTypeName);
+
+    if (bufOwnerRepoInfo != null) {
+      String[] parts = bufOwnerRepoInfo.split("/");
+      if (parts.length != 2) {
+        log.error("Cannot parse Buf owner and repo info from {}, make sure it is in the 'owner/repo' format",
+            bufOwnerRepoInfo);
+      } else {
+        bufOwner = parts[0];
+        bufRepo = parts[1];
+      }
+    } else {
+      String[] parts = fullyQualifiedTypeName.split("\\.");
+
+      if (parts.length == 0) {
+        log.warn("Cannot infer Buf repo name from type {}", fullyQualifiedTypeName);
+      } else {
+        bufRepo = parts[0];
+      }
+    }
+
+    log.info("Get descriptor from Buf {}/{}@{}", bufOwner, bufRepo, fullyQualifiedTypeName);
+
+    Descriptor descriptor = bufClient.getDescriptor(bufOwner, bufRepo, fullyQualifiedTypeName);
+
+    cachedDescriptor = new CachedDescriptor(currentDate, descriptor);
+    cachedMssageDescriptorMap.put(fullyQualifiedTypeName, cachedDescriptor);
+
+    return descriptor;
+  }
+
+  private String parse(byte[] value, Descriptor descriptor) throws IOException {
+    DynamicMessage protoMsg = DynamicMessage.parseFrom(
+        descriptor,
+        new ByteArrayInputStream(value));
+    byte[] jsonFromProto = ProtobufSchemaUtils.toJson(protoMsg);
+    return new String(jsonFromProto, utf8Charset);
+  }
+
   @Override
   public ProducerRecord<byte[], byte[]> serialize(String topic,
-                                                  @Nullable String key,
-                                                  @Nullable String data,
-                                                  @Nullable Integer partition) {
-    return this.schemaRegistryAwareRecordSerDe.serialize(topic, key, data, partition);
+      @Nullable String key,
+      @Nullable String data,
+      @Nullable Integer partition) {
+    if (data == null) {
+      return new ProducerRecord<>(topic, partition, Objects.requireNonNull(key).getBytes(), null);
+    }
+
+    ProtoSchema protoSchema = protoSchemaFromTopic(topic);
+    if (protoSchema == null) {
+      throw new RuntimeException("Could not infer proto schema from topic " + topic);
```

Comment:
RuntimeException is a pretty crappy exception, you should try to use a more specific exception. In this case I would suggest using `IllegalStateException`